File: benchmarks/sieve/sieve.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define LIMIT 10000000

int main(void) {
    char *flags = (char *)malloc(LIMIT + 1);
    memset(flags, 1, LIMIT + 1);
    flags[0] = 0;
    flags[1] = 0;
    for (long p = 2; p * p <= LIMIT; p++) {
        if (flags[p]) {
            for (long m = p * p; m <= LIMIT; m += p) {
                flags[m] = 0;
            }
        }
    }
    int count = 0;
    for (int i = 0; i <= LIMIT; i++) {
        if (flags[i]) count++;
    }
    printf("primes <= %d: %d\n", LIMIT, count);
    free(flags);
    return 0;
}

File: benchmarks/sieve/sieve.go
package main

import "fmt"

const LIMIT = 10000000

func main() {
	flags := make([]bool, LIMIT+1)
	for i := range flags {
		flags[i] = true
	}
	flags[0] = false
	flags[1] = false
	for p := 2; p*p <= LIMIT; p++ {
		if flags[p] {
			for m := p * p; m <= LIMIT; m += p {
				flags[m] = false
			}
		}
	}
	count := 0
	for i := 0; i <= LIMIT; i++ {
		if flags[i] {
			count++
		}
	}
	fmt.Printf("primes <= %d: %d\n", LIMIT, count)
}
